src/services/ComponentLoader.tsx

```diff
@@ -2,75 +2,210 @@ import React from 'react';
+    uri: string; 
+    type: string;
 }
+type uriOrString = string | TypedUri;
 
-const ComponentLoader = ({ component, data }: ComponentLoaderProps) => {
+type pathResult = uriOrString[] | uriOrString | undefined;
 
-    const getTreeValue = (fields) => {
-        let tree = { ...data };
+// `type: never` makes the type checker report an error if the case switch does not handle all types
+function checkUnknownComponent(type: never) {
+    console.error(`Type ${(type as ComponentConfig).type} is not handled!`);
+}
 
-        for (let i = 0, len = fields.length; i < len; i++) {
-            const key = fields[i];
+function isValue(obj: Value | Entity): obj is Value {
+    return obj.hasOwnProperty('value');
+}
 
-            if (!tree || !tree[key]) {
-                break;
-            }
+function walkPath(pathStr: string | undefined, formatters: FormatterConfig, entity: Entity): pathResult {
+    if (pathStr) {
+        const path = pathStr.split('.').map(segment => segment.split(':')[0]);
+        
+        return walkPathStep(path, formatters, entity);
+    }
+    return undefined;
+}
 
-            tree = tree[key];
+function walkPathStep(path: string[], formatters: FormatterConfig, entity: Entity): pathResult {
+    if (path.length === 0) {
+        return entity.uri;
+    }
+    let result = entity[path[0]];
+
+    if (!result) {
+        return undefined;
+    } else if (Array.isArray(result)) {
+        let retVal: uriOrString[] = [];
+        for (const item of result) {
+            if (isValue(item)) {
+                retVal.push(valueToString(item, formatters.concat([{
+                    type: 'http://timbuctoo.huygens.knaw.nl/datatypes/person-name',
+                    name: 'PERSON_NAMES'
+                }])));
+            } else {
+                const subResult = walkPathStep(path.slice(1), formatters, item);
+                if (subResult) {
+                    if (Array.isArray(subResult)) {
+                        retVal = retVal.concat(subResult);
+                    } else {
+                        retVal.push(subResult);
+                    }
+                }
+            }
         }
+        return retVal;
+    } else if (typeof result === 'string') {
+        console.log(entity, entity.__typename);
+        return {
+            uri: result,
+            type: entity.__typename
+        };
+    } else if (isValue(result)) {
+        return valueToString(result, formatters.concat([{
+            type: 'http://timbuctoo.huygens.knaw.nl/datatypes/person-name',
+            name: 'PERSON_NAMES'
+        }]));
+    } else {
+        return walkPathStep(path.slice(1), formatters, result);
+    }
+}
 
-        return tree.value ? tree.value : null;
-    };
-
-    const getValue = (values: ComponentValue) => {
-        if (!values) { return null; }
+function getValueOrLiteral(component: LeafComponentConfig | undefined, data: Entity): pathResult {
+    if (!component) {
+        return undefined;
```

Comment:
I always try to return 'null' instead of undefined. 

For two reasons mostly: 
- Stricter checking between null and undefined
-  if a bug occurs with 'undefined', I know for sure it's an unhandled exception

If you'd like to see that differently, I'll keep that in mind with my own coding ;-)